File: client/src/api/roles.ts
import { apiRequest } from "../services/api";

export interface Role {
  id: string;
  name: string;
  color: string | null;
  permissions: number;
  position: number;
  is_builtin: boolean;
  created_at: string;
  updated_at: string;
}

export interface ChannelPermissionOverride {
  channel_id: string;
  role_id: string;
  allow: number;
  deny: number;
}

export interface ListRolesResponse {
  roles: Role[];
}

export interface ListChannelOverridesResponse {
  overrides: ChannelPermissionOverride[];
}

export interface CreateRoleInput {
  name: string;
  color?: string | null;
  permissions: number;
  position?: number;
}

export interface UpdateRoleInput {
  name?: string;
  color?: string | null;
  permissions?: number;
  position?: number;
}

export async function listRoles(baseUrl: string, token: string): Promise<Role[]> {
  const response = await apiRequest<ListRolesResponse>(baseUrl, "/api/v1/roles", { token });
  return response.roles;
}

export async function createRole(
  baseUrl: string,
  token: string,
  input: CreateRoleInput,
): Promise<Role> {
  return apiRequest<Role>(baseUrl, "/api/v1/roles", {
    method: "POST",
    token,
    body: JSON.stringify(input),
  });
}

export async function updateRole(
  baseUrl: string,
  token: string,
  roleId: string,
  input: UpdateRoleInput,
): Promise<Role> {
  return apiRequest<Role>(baseUrl, `/api/v1/roles/${encodeURIComponent(roleId)}`, {
    method: "PATCH",
    token,
    body: JSON.stringify(input),
  });
}

export async function deleteRole(baseUrl: string, token: string, roleId: string): Promise<void> {
  await apiRequest<void>(baseUrl, `/api/v1/roles/${encodeURIComponent(roleId)}`, {
    method: "DELETE",
    token,
  });
}

export async function listChannelOverrides(
  baseUrl: string,
  token: string,
  channelId: string,
): Promise<ChannelPermissionOverride[]> {
  const response = await apiRequest<ListChannelOverridesResponse>(
    baseUrl,
    `/api/v1/channels/${encodeURIComponent(channelId)}/overrides`,
    { token },
  );
  return response.overrides;
}

export async function setChannelOverride(
  baseUrl: string,
  token: string,
  channelId: string,
  roleId: string,
  allow: number,
  deny: number,
): Promise<ChannelPermissionOverride> {
  return apiRequest<ChannelPermissionOverride>(
    baseUrl,
    `/api/v1/channels/${encodeURIComponent(channelId)}/overrides/${encodeURIComponent(roleId)}`,
    {
      method: "PUT",
      token,
      body: JSON.stringify({ allow, deny }),
    },
  );
}

export async function deleteChannelOverride(
  baseUrl: string,
  token: string,
  channelId: string,
  roleId: string,
): Promise<void> {
  await apiRequest<void>(
    baseUrl,
    `/api/v1/channels/${encodeURIComponent(channelId)}/overrides/${encodeURIComponent(roleId)}`,
    {
      method: "DELETE",
      token,
    },
  );
}

File: sdk/rust/src/rest/roles.rs
use crate::error::Result;
use crate::rest::models::{
    ChannelOverride, CreateRoleRequest, ListChannelOverridesResponse, ListRolesResponse, Role,
    SetChannelOverrideRequest, UpdateRoleRequest,
};
use crate::rest::Client;

pub struct RolesApi<'c> {
    client: &'c Client,
}

impl<'c> RolesApi<'c> {
    pub(crate) fn new(client: &'c Client) -> Self {
        Self { client }
    }

    pub async fn list(&self) -> Result<Vec<Role>> {
        let resp: ListRolesResponse = self.client.get("/api/v1/roles").await?;
        Ok(resp.roles)
    }

    pub async fn create(&self, req: &CreateRoleRequest) -> Result<Role> {
        self.client.post("/api/v1/roles", req).await
    }

    pub async fn update(&self, role_id: &str, req: &UpdateRoleRequest) -> Result<Role> {
        self.client
            .patch(&format!("/api/v1/roles/{role_id}"), req)
            .await
    }

    pub async fn delete(&self, role_id: &str) -> Result<()> {
        self.client
            .delete(&format!("/api/v1/roles/{role_id}"))
            .await
    }

    pub async fn add_member_role(&self, pubkey: &str, role_id: &str) -> Result<()> {
        self.client
            .put_empty(&format!("/api/v1/members/{pubkey}/roles/{role_id}"))
            .await
    }

    pub async fn remove_member_role(&self, pubkey: &str, role_id: &str) -> Result<()> {
        self.client
            .delete(&format!("/api/v1/members/{pubkey}/roles/{role_id}"))
            .await
    }

    pub async fn list_channel_overrides(
        &self,
        channel_id: &str,
    ) -> Result<Vec<ChannelOverride>> {
        let resp: ListChannelOverridesResponse = self
            .client
            .get(&format!("/api/v1/channels/{channel_id}/overrides"))
            .await?;
        Ok(resp.overrides)
    }

    pub async fn set_channel_override(
        &self,
        channel_id: &str,
        role_id: &str,
        req: &SetChannelOverrideRequest,
    ) -> Result<ChannelOverride> {
        self.client
            .put(
                &format!("/api/v1/channels/{channel_id}/overrides/{role_id}"),
                req,
            )
            .await
    }

    pub async fn delete_channel_override(
        &self,
        channel_id: &str,
        role_id: &str,
    ) -> Result<()> {
        self.client
            .delete(&format!(
                "/api/v1/channels/{channel_id}/overrides/{role_id}"
            ))
            .await
    }
}
